### PR TITLE
multi-vtep: set ovn-encap-ip on OVS Port when adding it to br-int

### PR DESCRIPTION
**Pull request by l8huang**
#### What this PR does and why is it needed

To utilize multi-VTEP, the `external_ids:ovn-encap-ip` of the VF's OVS Port
must be set. The VF's PCI device ID is then used to identify its corresponding
PF, which in turn is used to look up the `external_ids:ovn-pf-encap-ip-mapping`
to determine the encap-ip."

#### Special notes for reviewers
doc is in progress

#### How to verify it
Unit test cases are added to verify ``external_ids:ovn-encap-ip` is set on OVS Port.

### CODE REVIEW

**Comment by l8huang** [comment]
https://github.com/ovn-org/ovn-kubernetes/pull/4177 is required for this feature.

File: go-controller/pkg/cni/helper_linux.go [hunk 1]
```diff
@@ -368,9 +368,42 @@ func setupSriovInterface(netns ns.NetNS, containerID, ifName string, ifInfo *Pod
...
+	}
+
+	encapIP := encapIpMapping[uplinkRepName]
+	return encapIP, nil
+}
+
 // ConfigureOVS performs OVS configurations in order to set up Pod networking
 func ConfigureOVS(ctx context.Context, namespace, podName, hostIfaceName string,
-	ifInfo *PodInterfaceInfo, sandboxID string, getter PodInfoGetter) error {
+	ifInfo *PodInterfaceInfo, sandboxID string, deviceID string, getter PodInfoGetter) error {
```
Comment (cathy-zhou) [inline go-controller/pkg/cni/helper_linux.go:406]: can be: sandboxID, deviceID string,

**Review by cathy-zhou — COMMENTED** [review]
looks good to me, just a minor nit.

File: go-controller/pkg/cni/helper_linux.go (hunk 1 above)
Comment (l8huang) [reply go-controller/pkg/cni/helper_linux.go]: fixed.

**Comment by tssurya** [comment]
@l8huang / @girishmg : Is there an ETA for this, will it make it before Friday this week that's when we agreed for the 1.0.0 release cut, I hope we don't push it out further (but happy to get a reasonable ETA here else let's remove it from the roadmap milestone)

**Comment by l8huang** [comment]
@girishmg did you get a chance to review this PR?

File: docs/features/multiple-networks/multi-vtep.md [hunk 2]
```diff
@@ -1 +1,200 @@
-# MultiNetworkRails
\ No newline at end of file
+# Multi-VTEP
+
+## Introduction
+
+When a k8s cluser has multiple networks and k8s nodes have multiple SR-IOV
```
Comment (girishmg) [inline docs/features/multiple-networks/multi-vtep.md:5]: NIT: s/k8s/K8s/g (everywhere)

NIT: s/cluser/cluster/

File: docs/features/multiple-networks/multi-vtep.md [hunk 3]
```diff
@@ -1 +1,200 @@
...
+network adapters to provide different overlaying virtual networks over the
+physical networks, multple VTEP(Virtual Tunnel EndPoint) devices are
```
Comment (girishmg) [inline docs/features/multiple-networks/multi-vtep.md:7]: s/multiple/

File: docs/features/multiple-networks/multi-vtep.md [hunk 4]
```diff
@@ -1 +1,200 @@
...
+be routed exclusively through the second NIC, optimizing network performance
+and traffic segregation.
+
+![multi-vtep](../../images/multi-vtep.png)
+
+* vf<sup>R</sup> is VF representor.
+
+## How to enable this feature on an OVN Kubernetes cluster?
+
+On OVN Kubernetes side, no addtional configuration required to enable this feature.
```
Comment (girishmg) [inline docs/features/multiple-networks/multi-vtep.md:40]: NIT: additional

File: docs/features/multiple-networks/multi-vtep.md [hunk 5]
```diff
@@ -1 +1,200 @@
...
+
+This feature depends on a specific underlay network setup; it cannot be turned off
```
Comment (girishmg) [inline docs/features/multiple-networks/multi-vtep.md:42]: should this be `on`.

this feature cannot be turned on without an adequate underlay network configuration?

File: docs/features/multiple-networks/multi-vtep.md [hunk 6]
```diff
@@ -1 +1,200 @@
...
+```
+$ ovs-vsctl --timeout=15 --if-exists get Open_vSwitch . external_ids:ovn-encap-ip
+"10.0.0.1,10.0.0.2,10.0.0.3"
+
+$ ovs-vsctl --timeout=15 --if-exists get Open_vSwitch . external_ids:ovn-pf-encap-ip-mapping
+"enp1s0f0:10.0.0.1,enp2s0f0:10.0.0.2,enp3s0f0:10.0.0.3"
+```
+With above `external_ids:ovn-encap-ip` settings, OVN creates 3 tunnel ports for
+each encap-ip on remote node, that's, if remote node has 2 encap-ips in
+`external_ids:ovn-encap-ip`, 6 tunel ports will be created on the local node.
```
Comment (girishmg) [inline docs/features/multiple-networks/multi-vtep.md:140]: NIT: tunnel

File: docs/features/multiple-networks/multi-vtep.md [hunk 7]
```diff
@@ -1 +1,200 @@
...
+10.0.0.0/16 via 10.0.0.1 dev vtep1 proto static metric 802
+$ ip route list table 6083
+10.0.0.0/23 dev vtep2 scope link
+10.0.0.0/16 via 10.0.0.1 dev vtep2 proto static metric 806
+```
+
+## Known Limitations
+
+This feature only works with network adapters that support SR-IOV,
+e.g. NVIDIA CONNECTX-6 or NVIDIA BlueField-2 in NIC mode.
```
Comment (girishmg) [inline docs/features/multiple-networks/multi-vtep.md:200]: I think it works with any of the NICs, why only NVIDIA NICs?

**Review by girishmg — CHANGES_REQUESTED** [review]

File: docs/features/multiple-networks/multi-vtep.md (hunk 2 above)
Comment (l8huang) [reply docs/features/multiple-networks/multi-vtep.md]: done

File: docs/features/multiple-networks/multi-vtep.md (hunk 3 above)
Comment (l8huang) [reply docs/features/multiple-networks/multi-vtep.md]: done

File: docs/features/multiple-networks/multi-vtep.md (hunk 4 above)
Comment (l8huang) [reply docs/features/multiple-networks/multi-vtep.md]: done

File: docs/features/multiple-networks/multi-vtep.md (hunk 5 above)
Comment (l8huang) [reply docs/features/multiple-networks/multi-vtep.md]: you are right. 
done

File: docs/features/multiple-networks/multi-vtep.md (hunk 6 above)
Comment (l8huang) [reply docs/features/multiple-networks/multi-vtep.md]: done

File: docs/features/multiple-networks/multi-vtep.md (hunk 7 above)
Comment (l8huang) [reply docs/features/multiple-networks/multi-vtep.md:200]: NVIDIA NICs are examples(`e.g.`)

**Review by l8huang — COMMENTED** [review]
Thanks for your comments. I need to install spelling check in my editor.

**Review by girishmg — APPROVED** [review]
Thanks for addressing all my comments @l8huang ! 

LGTM

**Pull request merged**